testinfra/modules/blockdevice.py

```diff
@@ -0,0 +1,220 @@
+        None
+
+        """
+        if not self.is_block_device:
+            return None
+        return self._data['sector_size']
+
+    @property
+    def block_size(self):
+        """Return block size for the device in bytes or None.
+
+        >>> host.block_device("/dev/sda").block_szie
+        4096
+
+        >>> print(host.block_device("/dev/console").block_size)
+        None
+
+        """
+        if not self.is_block_device:
+            return None
+        return self._data['block_size']
+
+    @property
+    def start_sector(self):
+        """Return start sector of the device on the underlaying device.
+
+           Usually the value is zero for full devices and is non-zero
+           for partitions.
+           Return None if device is not a block device.
+
+        >>> host.block_device("/dev/sda1").start_sector
+        2048
+
+        >>> host.block_device("/dev/md0").start_sector
+        0
+
+        >>> print(host.block_device("/dev/console").start_sector)
+        None
+
+        """
+        if not self.is_block_device:
+            return None
+        return self._data['sector_size']
+
+    @property
+    def is_writable(self):
+        """Return True if device is writable (have no RO status), or None.
+
+        >>> host.block_device("/dev/sda").is_writable
+        True
+
+        >>> host.block_device("/dev/loop1").is_writable
+        False
+
+        >>> host.block_device("/dev/console").is_writable
+        None
+
+        """
+        if not self.is_block_device:
+            return None
+        mode = self._data['rw_mode']
+        if mode == 'rw':
+            return True
+        if mode == 'ro':
+            return False
+        raise ValueError('Unexpected value for rw: %s' % mode)
+
+    @property
+    def ra(self):
+        """Return Read Ahead for the device in 512-bytes sectors or None.
+
+        >>> host.block_device("/dev/sda").ra
+        256
+
+        >>> print(host.block_device("/dev/console").ra)
+        None
+        """
+        if not self.is_block_device:
+            return None
+        return self._data['read_ahead']
+
+    @property
+    def _data(self):
+        if self._data_cache is None:
+            self._data_cache = self._probe_device(self.device)
+        return self._data_cache
+
+    @classmethod
+    def get_module_class(cls, host):
+        if host.system_info.type == 'linux':
+            return LinuxBlockDevice
+        # if host.system_info.type.endswith("bsd"):
+        #     return BSDlockDevice
+        raise NotImplementedError
+
+    def __repr__(self):
+        return '<BlockDevice(path=%s)>' % self.device
+
+
+class LinuxBlockDevice(BlockDevice):
+
+    @classmethod
+    def _probe_device(cls, device):
+        HEADER = ['RO', 'RA', 'SSZ', 'BSZ', 'StartSec', 'Size', 'Device']
+        COMMAND = 'blockdev  --report %s'
+        run = cls(None).run
+        blockdev = run(COMMAND % device)
+        if blockdev.rc != 0:
```

Comment:
I'm not sure hidding the error is a good idea here.
Maybe just use `check_output()` instead and let the error go to the user.
This could avoid falsely successful tests.

Same remark for `is_block_device`, I think the module should only be used with a block device, if not just raise an error.

On my machine (debian stretch) I have:

```
$ sudo blockdev --report /dev/console ; echo $?
RO    RA   SSZ   BSZ   StartSec            Size   Device
blockdev: ioctl error on /dev/console
0
```
(I'd expect rc != 0 but...)


So I'd suggest to detect this by splitting the second line and expect exactly 7 items.

Reply:
Oh, that was my big fault. I tested that --report does not give a proper exit code, but forgot about this. I will rework this piece completely.

Big thanks for finding this out early.